src/main/java/org/jabref/logic/importer/fetcher/MergingIdBasedFetcher.java

```diff
@@ -0,0 +1,121 @@
+                          });
+        } catch (Exception exception) {
+            LOGGER.error("Error fetching entry with {} identifier: {}",
+                    field, identifier, exception);
+            return Optional.empty();
+        }
+    }
+
+    private boolean mergeBibEntry(BibEntry target, BibEntry source) {
+        boolean hasChanges = false;
+        for (Field field : source.getFields()) {
+            Optional<String> sourceValue = source.getField(field);
+            Optional<String> targetValue = target.getField(field);
+            if (sourceValue.isPresent() && !sourceValue.equals(targetValue)) {
+                target.setField(field, sourceValue.get());
+                hasChanges = true;
+            }
+        }
+        return hasChanges;
+    }
+
+    private Optional<IdBasedFetcher> getFetcherForField(Field field) {
+        return switch (field) {
+            case StandardField.DOI -> Optional.of(new DoiFetcher(importFormatPreferences));
+            case StandardField.ISBN -> Optional.of(new IsbnFetcher(importFormatPreferences));
+            case StandardField.EPRINT -> Optional.of(new IacrEprintFetcher(importFormatPreferences));
+            default -> Optional.empty();
+        };
+    }
+
+    private void logEntryIdentifiers(BibEntry entry) {
+        List<String> availableIds = Stream.of(StandardField.DOI, StandardField.ISBN, StandardField.EPRINT)
+                                          .flatMap(field -> entry.getField(field)
+                                                                 .map(value -> field +
+                                                                         ": " +
+                                                                         value)
+                                                                 .stream())
+                                          .collect(Collectors.toList());
+
+        String citationKey = entry.getCitationKey()
+                                  .map(key -> "citation key: " + key)
+                                  .orElse("no citation key");
+
+        LOGGER.debug("Processing entry with {} and identifiers: {}",
+                citationKey,
+                availableIds.isEmpty() ? "none" : String.join(", ", availableIds));
+    }
+
+    private void logEntryDetails(BibEntry entry, Field field, String identifier) {
+        StringBuilder details = new StringBuilder();
+        details.append(field).append(" identifier: ").append(identifier);
+
+        entry.getCitationKey().ifPresent(key -> details.append(", citation key: ").append(key));
+        entry.getField(StandardField.TITLE).ifPresent(title -> details.append(", title: ").append(title));
+        entry.getField(StandardField.AUTHOR).ifPresent(author -> details.append(", author: ").append(author));
+
+        LOGGER.debug("Entry details - {}", details);
+    }
+
+    public record FetcherResult(BibEntry originalEntry, BibEntry mergedEntry, boolean hasChanges) {
```

Comment:
Rename `originalEntry` to `entryFromLibrary`, because the entry comes from the BibDatabase, isn't it? (We are swithcing to use the term "library" instead of "database" for the local .bib file, therefore, we should use `Library` here)